src/mmda/predictors/hf_predictors/mention_predictor.py

```diff
@@ -101,84 +101,117 @@ def predict_page(self, page: Annotation, counter: Iterator[int], print_warnings:
         def has_label_id(lbls: List[int], want_label_id: int) -> bool:
             return any(lbl == want_label_id for lbl in lbls)
 
+        #  make list of word ids and list of label ids for each word
```

Comment:
can you help me understand what this means in this context? I see in the code above that `word` here is the joined symbols for a given token...but what does a word id or a word label id represent here? And what does it mean for a label id to be None?

Reply:
1. can you spell out for slow people like me why missing a word id from a previous batch would lead to an overlapping span group problem? 
2. do you think we could be making this same kind of batching error somewhere else in our code, or are the other overlapping span group errors unrelated? 
3. does this case argue that it would be bad to stop enforcing the no overlapping span group constraint?

Reply:
I'll try to explain as best as I can without honestly understanding the model process very well -- 
the `words` generated from token symbols are what we use as input to the model. Because the model can only take 512 tokens at a time, those `words` get split up into overlapping chunks (not sure how that overlapping is decided -- when debugging I saw the first page had a chunk of 140, then the rest were 512 each)
Those words get processed by the model and the output is lists of word id and corresponding label_id. I believe that within the model, other non-words get added in that help the model somehow -- "CLS" and "SEP" tokens, and those come back as parts that have no corresponding word_id (where we get `word_id` as `None`). 
The label_id is the number corresponding to the BILOU label the model assigned to it (when given only that chunk of tokens) that Yogi linked in a comment here: [this file, line 17](https://github.com/allenai/mmda/pull/252/files#diff-b0a7387d8d1575561368459c7d98159e6261f61809a514f0ab221bc31e69eb94L17)

1) This took me a long time to find the bug, I felt very slow to figure it out. What was happening was that each inner loop batch was checking the previous inner loop's last word to see if they matched, so that if they did, we didn't consider this word a new word, we considered it a word we've already seen and have a label for, so we add this new label to the list of `word_label_ids` (that lines up with `word_ids` -- now happening [here on line 118](https://github.com/allenai/mmda/pull/252/files#diff-b0a7387d8d1575561368459c7d98159e6261f61809a514f0ab221bc31e69eb94R118). 

When I was debugging with print statements, I could see that the last word_id in the word_ids was None when we hit word_id 433 for the second time (when we should have seen that the last word_id was also 433 in the last batch). It would reset to an empty word_list previously when going to the next outer-loop batch _after_ already adding the first word_id 433 and its span to the `ret` list of spangroups, so then we'd add it yet again (and end up with 2 spangroups that both have one of the same spans).

2) I don't know yet if we're making a similar error in VILA because I haven't looked there yet, I don't think we see overlapping spangroups in the other citation card models 

3) I do think this case argues that it would be bad to stop enforcing the no overlapping span group constraint because if we didn't have that in place, we wouldn't have noticed this bug easily. I think it would be difficult to find without manual human evaluation.

Reply:
I think 1 can be caused by a word being split between two chunks of tokens. Each chunk starts with CLS and ends with SEP, so the start and end word ids will be None. So if word 433 is split between two chunks of tokens then the second time it occurs it will be preceeded by None in the word id list